### ISSUE TITLE
Record vs RecordSet

### ISSUE THREAD

**Issue by rmsrosa** [beerjson/beerjson#41]
Both BeerXML 1.0 and the preliminary BeerXML 2.0.1 make a difference between a Record and a RecordSet.

I don't quite understand the json schema files, but, by looking at the examples in the test directory, I get the impression that everything is treated as a RecordSet in beerjson. Many filenames seem to indicate a single record but they include the plural mentioning the type of record, followed by a list object (although with a single element defining the specific record). Am I reading it wrong or is it really what you want? 

For example, instead of the file `HopWithRequiredFieldsOnly.json` reading

```{
  "beer_xml": {
    "version": 2.01,
    "hop_varieties": [
      {
        "name": "Cascade",
        "origin": "US",
        "alpha_acid_units": 5.5
      }
    ]
  }
}
```

couldn't it read

```{
  "beer_xml": {
    "version": 2.01,
    "hop": {
        "name": "Cascade",
        "origin": "US",
        "alpha_acid_units": 5.5
    }
  }
}
```

Is this allowed? 

If not, although it would require more work, wouldn't it be better to have a json-schema for single records and make the json-schema for record sets to have a list referencing the schema for a single record?

**Comment by krutilin** [comment]
I see. Checked names for records and yes hops.json and recipes.json could be renamed to the single form. Because they are items in the root array. This is an easy fix.

As about using single items instead of arrays. I am not sure how to make it better. Now we use arrays as it was described in the beerxml (need to check too).

Also, need to check is it possible to validate the object using only hops.json. If yes (and I suppose that yes) then you could construct your own definitions based on the schema.

**Comment by krutilin** [comment]
Well, an array with one item is equal to an individual record. And generally, we have set off something in the recipe. 
Renaming hops.json and recipes.json to hop.json and recipe.json looks good for me. 

With other need feedback from the community.

**Comment by rmsrosa** [comment]
Sorry for asking, but I don't known JS very much. I understand a JS value can be an array or a single item. And I believe a json schema can be configure to allow or not arrays, right? But I also assume that the current beerjson schema does not enforce that. Hence would both codes

```
{
  "beer_xml": {
    "version": 2.01,
    "hop_varieties": [
      {
        "name": "Cascade",
        "origin": "US",
        "alpha_acid_units": 5.5
      }
    ]
  }
}
```

and

```
{
  "beer_xml": {
    "version": 2.01,
    "hop_varieties": {
        "name": "Cascade",
        "origin": "US",
        "alpha_acid_units": 5.5
    }
  }
}
```

be allowed and pass the validation? 

And I think I don't understand your proposal of simply renaming hops.json and recipes.json to hop.json and recipe.json. Either those schemas are good for one or several hops and the plural should be kept in the name, or the schemas are changed to allowing only one hop and only one recipe and then the changes to singular hop.json and recipe.json make sense.

Sorry, but I am  a bit lost.

**Comment by krutilin** [comment]
Well, it is possible to make experiments in tests and check it. By default constructions without array will fail validation. We could rewrite root schema in this way:
```json
            "recipes": {
              "description":"Records containing bla bla bla",
              "oneOf": [
                {
                  "type": "array",
                  "items": {
                    "$ref": "recipes.json#/definitions/RecipeType"
                  }
                },
                {
                  "type": "object",
                  "$ref": "recipes.json#/definitions/RecipeType"
                }
              ]
            },
```
to archive validation for arrays and single objects. But I can't understand key reason for it. Brewing software can handle arrays, single objects and arrays with one item without any troubles.

**Comment by rmsrosa** [comment]
It is not really necessary. I agree that brewing softwares can handle either way, and, from a standpoint of exchanging recipes and databases, arrays are more than adequate. 

There are only minor reasons for changing it: for similitude with beerxml, which allows single items;  for completeness of the schema; and to eliminate unnecessary arrays when, e.g., writing a recipe with a single mash, and/or a single boil, and/or a single malt, and/or a single hop addition, and/or a single yeast. In theses cases, the recipe will have one or more unnecessary arrays. But, as I said, it is no big deal, we can leave it as it is, if everyone is comfortable with that.

In case we leave it as it is, I think at least that the documentation deserves a remark saying that, contrarily to the beerxml schema, the beerjson schema requires an array even in case of single items, i.e. the json for a single item is an array with a single element. Something along these lines.

**Comment by krutilin** [comment]
@rmsrosa I recommend you to play with tests and schema (see docs to this repo). Because validation is a powerfull thing. When you apply the schema to the incorrect json, you will receive message why it is incorrect. Like this one:
`[ { keyword: 'type',
    dataPath: '.beer_xml.recipes',
    schemaPath: '#/properties/beer_xml/allOf/0/properties/recipes/type',
    params: { type: 'array' },
    message: 'should be array' } ]`

And the question to @beerjson/beerjson-wg should we support single items in root beerjson schema?

**Comment by rmsrosa** [comment]
Yeah, I know, I should take some time to learn about this npm, install it and learn how to use this schema stuff.

**Comment by rmsrosa** [comment]
I managed to install the latest Node.js, which already comes with npm, and was able to locally run `npm test`. Pretty cool. I played a little bit with it, purposely making some erros to see how it goes. But I still need to understand the syntax of json schema to be able to modify it without making a mess. For instance, I tried to adapt in hops.json the modification above that @krutilin suggested on recipes.json to work with either an object or an array but I failed. 

Anyway, if I succeed in understanding json-schema, then I might try addressing some of the issues, instead of only making remarks. 

I will be following https://spacetelescope.github.io/understanding-json-schema/index.html to learn more about json schemas. It looks nice.
